### ISSUE TITLE
Byzantine "Marching" Problem: Keeping Nodes In The Same Timeframe

### ISSUE THREAD

**Issue by 00hello** [grasslandnetwork/node_lite#6]
I've been working on a fault tolerant metronome algorithm for the Full Node implementation to ensure that no matter what happens every node can stay in sync with the others as to the time frame they're all capturing even if there is a problem with network's communication. And to ensure that as many nodes as possible reach consensus and make it into the next block.

The problem is like this. Imagine you have random group of strangers. And you have to get them to walk in lockstep so they're all marching in a parallel direction and with the same rhythm no matter where they are on earth. But because no one's in the same location they can't see each other and even communicating with one another can be very unreliable. And of course they all have varying stride lengths. So to solve this problem for our case, we need to find a way to make sure:

1. Each node is somehow synchronized with all the other nodes around the world as to the 10 minute timespan they're all recording so they all start the next sequence at the same time.
2. That every node that makes it into a block has expended the same total amount of computation (the difficulty target). Because if they're all processing a different number of frames, then both the blocks and the Script language we use for validating blocks require much more complexity to account for paying them each differently. And complexity is the death of security. So that's not an option.
3. That as many nodes as possible have the computational power to keep up with the rest of the group 
4. That a balance is struck between the minimum acceptable computational power and the security needs of the network w.r.t. preventing fake data
5. The probability that every node in a block has captured data from the real world grows higher and higher.
6. That it's easy for any node to verify that a particular node expended the same computation they said they did (proof-of-work)
7. That they all agree on a particular history of the world and that as many nodes as possible have a sufficient amount of time to reach consensus as to which history they're going with
8. That what they're capturing is live. And we know the arrow of time in the real world only moves in one direction

So my solution to ensuring the above is to have the aborist's algorithm that puts nodes into the next block check ...

1. That every node processed the exact same number of frames over that block's 10 minute sequence. 
2. That the RGB values in the four corners of frame N prior to being processed by the artificial neural network were replaced with the first 12 bytes of the hash of frame N-1.
3. But if that frame is the first frame of the next 10 minute sequence, the values in each of its four corners should instead be the first 12 bytes from the block hash of the second to last block. (Those four corners aren't enough to distort the objects the neural network detects in the frame but it will significantly change the hash of the network's hidden layers)
4. That every node in a block agrees as to which block is the second to last block (there isn't enough time if it was the very last block). So their first frame must all have the same bytes in their four corners
5. That the solar azimuth (shadows) detected by the neural network of every node in a block corresponds correctly to the spot on they're earth they're saying they're looking at and the time of day for that block.

Now, as far as I can tell, this algorithm accomplishes all of those requirements by creating a Byzantine fault tolerant "drummer-boy" or metronome to help the nodes "keep time", so to speak. The hashes in the four corners prevents nodes from parallelizing the frame processing which makes it so that every node "marches" in the same direction following the arrow of time and that nodes with differing computational power don't fall out of step with each other which would mean a complete loss of consensus. And it works even if there is network latency or a disruption in communication lasting several blocks.

**Comment by dstorrs** [comment]
Interesting problem, and all the kudos for finding a cool solution.

As to the solution itself: 1-4 seem reasonable.  5 sounds good but I'd be wary of implementation.  Seems like there's a lot of potential failure modes (local terrain, artificial light sources, etc) that would cause the azimuth to be mis-estimated.

**Comment by 00hello** [comment]
Yeah. Solar azimuth detection isn't always accurate. But with 6000 frames every ten minutes (10 FPS) for each node, you're looking at the average across all frames and all nodes. Any node outside say one or two standard deviations doesn't get in the block.

And since the blocks will be produced live, every ten minutes, it helps the nodes reach consensus for each block's time of day. That lets them use the detected solar azimuth to have a rough validation for every node's longitude as well as the latitude.
